src/mapping_utils.py

```diff
@@ -41,34 +44,43 @@ def plot_frequency_map(cls, geo_data, admin_id_column, frequencies, label_positi
-
-        # Plot the choropleth map.
-        # Frequencies are classed using the Fisher-Jenks method, a standard GIS algorithm for choropleth classification.
+        # Class the frequencies using the Fisher-Jenks method, a standard GIS algorithm for choropleth classification.
         # Using this method prevents a region with a vastly higher frequency than the others (e.g. a capital city)
         # from using up all of the colour range, as would happen with a linear scale.
-        unique_frequencies = {f for f in frequencies.values() if f != 0}
-        number_of_classes = min(5, len(unique_frequencies))
+        # Ignores zeros when classing, so that 0s are not included in the same class as other lower numbers, then adds
+        # the 0 back in when converting from classes to bin edges.
+        frequencies_to_class = [f for f in frequencies.values() if f != 0]
+        number_of_classes = min(5, len(set(frequencies_to_class)))
+        bin_edges = [0]
         if number_of_classes > 0:
-            geo_data.plot(ax=ax,
-                          column="Frequency", cmap=cls.AVF_COLOR_MAP,
-                          scheme="fisher_jenks", k=number_of_classes,
-                          linewidth=0.1, edgecolor="black",
-                          missing_kwds={"edgecolor": "black", "facecolor": "white"})
-        else:
-            # Special-case plotter for when all frequencies are 0, because geopandas crashes otherwise.
-            geo_data.plot(ax=plt.gca(), linewidth=0.1, edgecolor="black", facecolor="white")
+            bin_edges.extend(FisherJenks(np.array(frequencies_to_class), k=number_of_classes).bins)
+
+        # Get the color for each region by searching for the appropriate bin for each frequency.
+        colors = []
+        for i, admin_region in geo_data.iterrows():
+            frequency = frequencies[admin_region[admin_id_column]]
+            bin_id = [i for i, b in enumerate(bin_edges) if b >= frequency][0]  # Index of first bin edge >= frequency
+            colors.append(cls.AVF_COLOR_MAP(0 if bin_id == 0 else float(bin_id) / number_of_classes))
+
+        # Plot the choropleth map.
+        ax = geo_data.plot(ax=ax, color=colors, linewidth=0.1, edgecolor="black")
         plt.axis("off")
 
+        # Add the choropleth legend.
+        legend_elements = [
+            Patch(label="0", facecolor=cls.AVF_COLOR_MAP(0), linewidth=0.1, edgecolor="black")
+        ]
+        for bin_id in range(1, len(bin_edges)):
+            range_min = bin_edges[bin_id - 1] + 1
+            range_max = bin_edges[bin_id]
+            legend_elements.append(Patch(
+                label=range_min if range_min == range_max else f"{range_min} - {range_max}",
+                facecolor=cls.AVF_COLOR_MAP(float(bin_id) / number_of_classes),
+                linewidth=0.1, edgecolor="black"
+            ))
+        ax.legend(handles=legend_elements, title="Participants", title_fontsize=6, loc="lower right",
```

Comment:
```suggestion
        ax.legend(handles=legend_elements, title="No. of participants", title_fontsize=6, loc="lower right",
```

Reply:
Unfortunately this doesn't fit.
We need to do some more work to be able to have better control over placement of the legend, both for this and for future maps where a corner doesn't work. I will leave for now but add an issue.

Reply:
https://github.com/AfricasVoices/Project-COVID19/issues/47